OpenRA.Mods.Common/Effects/LineImpactProjectile.cs

```diff
@@ -0,0 +1,261 @@
+/*
+ * Copyright 2007-2014 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation. For more information,
+ * see COPYING.
+ */
+#endregion
+
+using System;
+using System.Collections.Generic;
+using System.Drawing;
+using System.Linq;
+using OpenRA.Effects;
+using OpenRA.GameRules;
+using OpenRA.Graphics;
+using OpenRA.Mods.Common.Graphics;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Effects
+{
+	public class LineImpactProjectileInfo : IProjectileInfo
+	{
+		[Desc("Projectile speed in WDist / tick,", "two values indicate variable velocity")]
+		public readonly WDist[] Speed = { new WDist(17) };
+
+		[Desc("The number of ticks between the beam", "causing warhead impacts in its area of effect")]
+		public readonly int TicksBetweenImpacts = 3;
+
+		[Desc("The distance between applying the", "warhead effects")]
+		public readonly WDist RangeBetweenImpacts = new WDist(16);
+
+		[Desc("How far beyond the target the", "projectile keeps on travelling")]
+		public readonly WDist BeyondTargetRange = new WDist(128);
+
+		[Desc("How close the target must the", "projectile be to apply weapon effects")]
+		public readonly WDist MaxTargetRange = new WDist(128);
+
+		[Desc("How far away from its source must", "the projectile be before applying weapon effects")]
+		public readonly WDist MinRangeSource = new WDist(256);
+
+		[Desc("The maximum reduction to warhead effects", "1000 = 100% falloff = 0% damage")]
+		public readonly int FalloffMaxReduction = 750;
+
+		[Desc("The distance before the falloff starts")]
+		public readonly WDist FalloffStartDistance = new WDist(128);
+
+		[Desc("The distance before the falloff ends")]
+		public readonly WDist FalloffEndDistance = new WDist(4096);
+
+		[Desc("Maximum offset at the maximum range")]
+		public readonly WDist Inaccuracy = WDist.Zero;
+
+		[Desc("Check for whether an actor with Wall: trait blocks fire")]
+		public readonly bool High = false;
+
+		[Desc("The width of the beam projectile")]
+		public readonly int BeamWidth = 2;
+
+		[Desc("The duration of the beam projectile", "effectively determines the length of the beam")]
+		public readonly int BeamFireDuration = 10;
+
+		[Desc("Beam color in (A,)R,G,B.")]
+		public readonly Color Color = Color.Red;
+
+		[Desc("Beam color is the player's color")]
+		public readonly bool UsePlayerColor = false;
+
+		public IEffect Create(ProjectileArgs args)
+		{
+			var c = UsePlayerColor ? args.SourceActor.Owner.Color.RGB : Color;
+			return new LineImpactProjectile(this, args, c);
+		}
+	}
+
+	public class LineImpactProjectile : IEffect, ISync
+	{
+		readonly LineImpactProjectileInfo info;
+		readonly ProjectileArgs args;
+		[Sync] readonly WDist speed;
+
+		Color color;
+		WAngle angle;
+		WVec offSet;
+
+		[Sync] WPos headPos;
+		[Sync] WPos tailPos;
+		[Sync] WPos target;
+		[Sync] int length;
```

Comment:
`[Sync]`ing everything hurts performance.  You should only sync parameters that need it to uniquely identify desyncs (e.g. length is uniquely determined by `headPos`, `target`, and `speed`, which are also synced).